.github/workflows/smoketest.yml

```diff
@@ -119,38 +140,44 @@ jobs:
         echo "Action     = ${{ github.action }}"
         echo "Event      = ${{ github.event_name }}"
         echo "Actor      = ${{ github.actor }}"
+        echo "Ref        = ${{ github.ref }}"
+        echo "Sha        = ${{ github.sha }}"
         echo "github.repository = ${{ github.repository }}"
         echo "repository = ${{ steps.vars.outputs.repository }}"
         echo "repository_owner = ${{ github.repository_owner }}"
         echo "github.workspace = ${{ github.workspace }}"
         echo "runner.workspace = ${{ runner.workspace }}"
         echo "github.event.pull_request.head.repo.owner.login = ${{ github.event.pull_request.head.repo.owner.login }}"
 
-        if [[ "${{ github.event_name }}" = 'pull_request' ]]; then \
-          echo "PR base_ref = ${{ github.base_ref }}"; \
-          echo "PR head_ref = ${{ github.head_ref }}"; \
-          echo "PR base SHA = ${{ github.event.pull_request.base.sha }}"; \
+        if [[ "${{ github.event_name }}" = 'pull_request' ]]
+        then
+          echo "PR base_ref = ${{ github.base_ref }}"
+          echo "PR head_ref = ${{ github.head_ref }}"
+          echo "PR base SHA = ${{ github.event.pull_request.base.sha }}"
         fi
-        echo restore-key1: hpccbuild-${{ env.cacheversion }}-${{ steps.vars.outputs.base_ref }}
-        echo restore-key2: ${{ steps.vars.outputs.base-majorminor-restore-key }}
-        echo restore-key3: ${{ steps.vars.outputs.base-majoronly-restore-key }}
-        echo restore-key4: hpccbuild-${{ env.cacheversion }}-
+        echo "restore-key1: hpccbuild-${{ env.cacheversion }}-${{ steps.vars.outputs.base_ref }}"
+        echo "restore-key2: ${{ steps.vars.outputs.base-majorminor-restore-key }}"
+        echo "restore-key3: ${{ steps.vars.outputs.base-majoronly-restore-key }}"
+        echo "restore-key4: hpccbuild-${{ env.cacheversion }}-"
 
     # NB: actions/cache does not set cache-hit to true if restore-keys used.
     # So here we check if a cache was restored, and set a separate variable 'cache-exists' if it does.
     - name: Check cache
-      if: steps.cache.outputs.cache-hit != 'true'
       id: check-cache
+      if: steps.cache.outputs.cache-hit != 'true'
       run: |
-        if [[ -n "${{ secrets.SMOKETEST_SKIP_CACHE }}" ]]; then \
-          echo "Cache found, but ignored because SMOKETEST_SKIP_CACHE set" ;\
-          rm -rf build-cache src-cache; \
+        if [[ -n "${{ secrets.SMOKETEST_SKIP_CACHE }}" ]]
+        then
+          echo "Cache found, but ignored because SMOKETEST_SKIP_CACHE set"
+          rm -rf build-cache merge-patch src-cache
         else
-          echo ::set-output name=cache-exists::$(if [[ -d build-cache ]]; then echo true; else echo false; fi) ;\
-          if [[ -d build-cache ]]; then \
-            stat build-cache src-cache; \
-            du -sh build-cache src-cache; \
-          fi ;\
+          if [[ -d build-cache ]]
```

Comment:
The actions/cache "cache-hit" flag is only set on an exact match, which we are generally _not_ interested in, because it suggests a rerun of this action that has succeeded before, and there should be nothing to do.

We are interested in the inexact match (cache restored via restored-keys), which means a close/closest match was found - it is on that basis that it uses that previous build, to avoid a complete rebuild.
We know which one by virtue of the 'cache-sha' file from the restored cache item + the 'merge-patch' - which are used to restore the same source state - to match the restored built binaries (ensuring they are not rebuilt).

Reply:
> Moreover if there is no matching key nor matching restore-key what this means:
>
>> 3. If the job completes successfully, the action creates a new cache with the contents of the path directory.
>When this action can be successful?

It is successful either via a cache-miss, where it does a full build, no reliance on cache objects, OR
on a 'cache-exists' where it follows the cache-hit step and uses the cache item to perform a partial build.

It's also successful on a full cache-hit, but it doesn't really do anything in that case.